kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1180 @@
 //! Some utilities for working with arrow data types
 
-use std::sync::Arc;
+use std::{collections::HashSet, sync::Arc};
 
-use crate::{schema::SchemaRef, utils::require, DeltaResult, Error};
+use crate::{
+    schema::{DataType, PrimitiveType, Schema, SchemaRef, StructField, StructType},
+    utils::require,
+    DeltaResult, Error,
+};
 
-use arrow_array::RecordBatch;
-use arrow_schema::{Schema as ArrowSchema, SchemaRef as ArrowSchemaRef};
+use arrow_array::{
+    cast::AsArray, new_null_array, Array as ArrowArray, GenericListArray, OffsetSizeTrait,
+    StructArray,
+};
+use arrow_schema::{
+    DataType as ArrowDataType, Field as ArrowField, FieldRef as ArrowFieldRef, Fields,
+    SchemaRef as ArrowSchemaRef,
+};
+use itertools::Itertools;
 use parquet::{arrow::ProjectionMask, schema::types::SchemaDescriptor};
+use tracing::debug;
 
-/// Get the indicies in `parquet_schema` of the specified columns in `requested_schema`. This
-/// returns a tuples of (mask_indicies: Vec<parquet_schema_index>, reorder_indicies:
-/// Vec<requested_index>). `mask_indicies` is used for generating the mask for reading from the
+fn make_arrow_error(s: String) -> Error {
+    Error::Arrow(arrow_schema::ArrowError::InvalidArgumentError(s))
+}
+
+/// Ensure a kernel data type matches an arrow data type. This only ensures that the actual "type"
+/// is the same, but does so recursively into structs, and ensures lists and maps have the correct
+/// associated types as well. This returns an `Ok(())` if the types are compatible, or an error if
+/// the types do not match. If there is a `struct` type included, we only ensure that the named
+/// fields that the kernel is asking for exist, and that for those fields the types
+/// match. Un-selected fields are ignored.
+pub(crate) fn ensure_data_types(
+    kernel_type: &DataType,
+    arrow_type: &ArrowDataType,
+) -> DeltaResult<()> {
+    match (kernel_type, arrow_type) {
+        (DataType::Primitive(PrimitiveType::Timestamp), ArrowDataType::Timestamp(_, _))
+        | (DataType::Primitive(PrimitiveType::TimestampNtz), ArrowDataType::Timestamp(_, _)) => {
+            // We assume that any timestamp data read from a delta table is correctly written in
```

Comment:
note: this is... questionable. we should not do this long term, but i need to figure out what's going on on mac such that things read incorrectly, and then if it really is reading in nanoseconds for some reason, we'd need to do some kind of conversion on the fly (maybe).

Regardless, not something I want to tackle in this PR.

Note that the file in question is `acceptance/tests/dat/out/reader_tests/generated/all_primitive_types/delta/part-00000-a31beaf2-ba0b-4ad6-a402-e5867ba52c91-c000.snappy.parquet` and it really does seem to have things stored in microseconds with UTC

<details>
  <summary>Running `parquet-tools summary` on the file</summary>

```
$ parquet-tools inspect part-00000-a31beaf2-ba0b-4ad6-a402-e5867ba52c91-c000.snappy.parquet
############ file meta data ############
created_by: parquet-mr version 1.13.1 (build db4183109d5b734ec5930d870cdae161e408ddba)
num_columns: 12
num_rows: 5
num_row_groups: 1
format_version: 1.0
serialized_size: 2189


############ Columns ############
utf8
int64
int32
int16
int8
float32
float64
bool
binary
decimal
date32
timestamp

############ Column(utf8) ############
name: utf8
path: utf8
max_definition_level: 1
max_repetition_level: 0
physical_type: BYTE_ARRAY
logical_type: String
converted_type (legacy): UTF8
compression: SNAPPY (space_saved: -2%)

...[snip]...

############ Column(timestamp) ############
name: timestamp
path: timestamp
max_definition_level: 1
max_repetition_level: 0
physical_type: INT64
logical_type: Timestamp(isAdjustedToUTC=true, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false)
converted_type (legacy): TIMESTAMP_MICROS
compression: SNAPPY (space_saved: 0%)
```

</details>

But the mac tests fail if I don't use this hack: https://github.com/delta-incubator/delta-kernel-rs/actions/runs/9882874132/job/27296554771 complaining  `"Incorrect datatype. Expected Timestamp(Microsecond, Some(\"UTC\")), got Timestamp(Nanosecond, None)"`

Reply:
Ahh, nm, this is an issue with DAT. My local version works because the file there has the correct format, but the one the test runners pull is incorrect. I will figure out how to fix that and revert this hack.

Reply:
Okay, this turned out to be something that's fixed in the newest version of `dat`, so I've reverted the hack, and bumped the version for our acceptance tests.

Created #279 to follow up on this